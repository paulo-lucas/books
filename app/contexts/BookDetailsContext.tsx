import React, {
  createContext,
  PropsWithChildren,
  useCallback,
  useEffect,
  useState,
} from 'react';
import { RouteProp, useRoute, useNavigation } from '@react-navigation/native';
import type { BookDetailsContextData } from '@app/interfaces/bookDetailsContext';
import type { Volume } from '@app/interfaces/volume';
import { useFavorites } from '@app/hooks';
import type { RootStackParamList } from '@app/router/rootStackParams';
import { showVolume } from '@app/services';
import { Alert } from 'react-native';

export const BookDetailsContext = createContext({} as BookDetailsContextData);

export const BookDetailsProvider: React.FC<PropsWithChildren> = ({
  children,
}) => {
  const navigation = useNavigation();
  const {
    params: { id },
  } = useRoute<RouteProp<RootStackParamList, 'Details'>>();

  const [book, setBook] = useState<Volume>();
  const [loading, setLoading] = useState<boolean>(true);
  const { isFavorite, toggleFavorite } = useFavorites();

  const fetchBook = useCallback(async () => {
    if (!id) {
      return navigation.goBack();
    }

    try {
      setLoading(true);
      const { data } = await showVolume(id);
      setBook(data);
      setLoading(false);
    } catch (err) {
      Alert.alert('Could not retrieve book.');
      navigation.goBack();
    }
  }, [id, navigation]);

  useEffect(() => {
    fetchBook();
  }, [fetchBook]);

  const contextData: BookDetailsContextData = {
    book,
    loading,
    isFavorite: () => isFavorite(id),
    toggleFavorite: () => toggleFavorite(id),
  };

  return (
    <BookDetailsContext.Provider value={contextData}>
      {children}
    </BookDetailsContext.Provider>
  );
};
