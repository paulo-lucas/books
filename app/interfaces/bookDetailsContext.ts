import type { Volume } from './volume';

export interface BooksDetailsContextData {
  book: Volume;
  toggleFavorite(): void;
  isFavorite(): boolean;
}
