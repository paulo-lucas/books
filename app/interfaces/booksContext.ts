import type { Volume } from './volume';

export interface BooksState {
  total: number;
  itemsPerPage: number;
  page: number;
  data: Array<Volume>;
}

export interface BooksContextData {
  searchText: string;
  onChangeSearch(text: string): void;
  books: BooksState;
  refreshing: boolean;
  toggleOrderBy(): void;
  toggleFavoriteFilter(): void;
  toggleFavorite(isbn: string): void;
  isFavorite(favorite: string): boolean;
  onClearSearch(): void;
  refreshBooks(page?: number): void;
}

export interface UpdatePayload {
  type: 'update';
  page?: number;
  total: number;
  data: Array<Volume>;
}

export interface ClearPayload {
  type: 'clear';
}
