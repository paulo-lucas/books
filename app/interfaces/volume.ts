export interface VolumeSearchRequest {
  q: VolumeSearchQuery;
  startIndex: number;
  maxResults: number;
  orderBy: 'newest' | 'relevance';
}

export interface VolumeSearchResponse {
  kind: string;
  totalItems: number;
  items: Array<Volume>;
}

export interface VolumeSearchQuery {
  search?: string;
  title?: string;
  author?: string;
  publisher?: string;
  isbn?: Array<string>;
}

export interface VolumeInfo {
  title: string;
  authors: Array<String>;
  publisher: string;
  publishedDate: string;
  description: string;
  pageCount: number;
  categories: Array<string>;
  maturityRating: string;
  imageLinks: { smallThumbnail: string; thumbnail: string };
  language: string;
  previewLink: string;
  infoLink: string;
  canonicalVolumeLink: string;
}

export interface SaleInfo {
  country: string;
  saleability: string;
  isEbook: boolean;
}

export interface SearchInfo {
  textSnippet: string;
}

export interface Volume {
  kind: string;
  id: string;
  selfLink: string;
  volumeInfo: VolumeInfo;
  saleInfo: SaleInfo;
  accessInfo: any;
  searchInfo: SearchInfo;
}
