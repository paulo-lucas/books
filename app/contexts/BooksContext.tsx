import React, {
  createContext,
  useState,
  useReducer,
  PropsWithChildren,
  useEffect,
} from 'react';
import { BooksContextData } from '@app/interfaces/booksContext';
import { booksReducer, booksInitialState } from '@app/states/BooksReducer';
import { useDebounce } from '@app/hooks';
import { searchVolumes, storeFavorites, getFavorites } from '@app/services';
import { Alert } from 'react-native';
import { useCallback } from 'react';

export const BooksContext = createContext({} as BooksContextData);

export const BooksProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [books, dispatch] = useReducer(booksReducer, booksInitialState);
  const [searchText, setSearchText] = useState<string>('');
  const [refreshing, setRefreshing] = useState<boolean>(false);
  const [orderBy, setOrderBy] = useState<'relevance' | 'newest'>('relevance');
  const [favorites, setFavorites] = useState<Array<string>>([]);
  const [filterByFavorites, setFilterByFavorites] = useState<boolean>(false);

  const fetchFavorites = useCallback(async () => {
    const storageFavorites = await getFavorites();
    setFavorites(storageFavorites);
  }, []);

  const onClearSearch = () => dispatch({ type: 'clear' });

  const toggleFavoriteFilter = () => setFilterByFavorites(!filterByFavorites);

  const toggleOrderBy = () =>
    setOrderBy(orderBy === 'newest' ? 'relevance' : 'newest');

  const isFavorite = (isbn: string) => favorites.includes(isbn);

  const toggleFavorite = (isbn: string) => {
    let updatedFavorites;

    if (favorites.includes(isbn)) {
      updatedFavorites = favorites.filter(fav => fav !== isbn);
    } else {
      updatedFavorites = [...favorites, isbn];
    }

    setFavorites(updatedFavorites);
    storeFavorites(updatedFavorites);
  };

  const fetchBooks = useCallback(
    async (page: number, searchFromDebounce?: string) => {
      if (!searchText && !searchFromDebounce) {
        return onClearSearch();
      }

      const search = searchFromDebounce ?? searchText;
      const isbn = filterByFavorites ? favorites : [];
      const q = { search, isbn };
      const { itemsPerPage: maxResults } = books;
      const startIndex = page * maxResults;

      try {
        setRefreshing(true);
        const { data } = await searchVolumes({
          q,
          startIndex,
          maxResults,
          orderBy,
        });

        dispatch({
          type: 'update',
          total: data.totalItems,
          data: data.items ?? [],
        });

        setRefreshing(false);
      } catch (err) {
        Alert.alert('Could not retrieve books.');
      }
    },
    [books, favorites, filterByFavorites, orderBy, searchText],
  );

  const debounceSearch = useDebounce(fetchBooks, 1000);

  const onChangeSearch = (text: string) => {
    setSearchText(text);
    debounceSearch(0, text);
  };

  const refreshBooks = (page: number = 0) => {
    fetchBooks(page);
  };

  useEffect(() => {
    fetchFavorites();
  }, [fetchFavorites]);

  const contextData: BooksContextData = {
    books,
    refreshBooks,
    refreshing,
    searchText,
    onChangeSearch,
    onClearSearch,
    toggleOrderBy,
    toggleFavoriteFilter,
    isFavorite,
    toggleFavorite,
  };

  return (
    <BooksContext.Provider value={contextData}>
      {children}
    </BooksContext.Provider>
  );
};
