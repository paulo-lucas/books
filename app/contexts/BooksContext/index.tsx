import React, {
  createContext,
  useState,
  useReducer,
  PropsWithChildren,
  useEffect,
} from 'react';
import { BooksContextData } from '@app/interfaces/booksContext';
import { booksReducer, booksInitialState } from '@app/states/BooksReducer';
import { useDebounce } from '@app/hooks';
import { searchVolumes, storeFavorites, getFavorites } from '@app/services';
import { Alert } from 'react-native';
import { useCallback } from 'react';
import { VolumeSearchRequest } from '@app/interfaces/volume';

export const BooksContext = createContext({} as BooksContextData);

export const BooksProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [books, dispatch] = useReducer(booksReducer, booksInitialState);
  const [searchText, setSearchText] = useState<string>('');
  const [searchTextDebounce, setSearchTextDebounce] = useState<string>('');
  const [refreshing, setRefreshing] = useState<boolean>(false);
  const [orderBy, setOrderBy] = useState<'relevance' | 'newest'>('relevance');
  const [favorites, setFavorites] = useState<Array<string>>([]);
  const [filterByFavorites, setFilterByFavorites] = useState<boolean>(false);

  const fetchFavorites = useCallback(async () => {
    const favoritesFromStorage = await getFavorites();
    setFavorites(favoritesFromStorage);
  }, []);

  // Handle every request to list books, based on states
  const fetchBooks = useCallback(async () => {
    if (!searchTextDebounce && (!filterByFavorites || !favorites.length)) {
      return onClearSearch();
    }

    const identifier = filterByFavorites ? favorites : [];
    const q = { search: searchTextDebounce, identifier };
    const { itemsPerPage: maxResults, page } = books;
    const startIndex = page * maxResults;

    try {
      const request: VolumeSearchRequest = {
        q,
        startIndex,
        maxResults,
        orderBy,
        filter: 'partial',
        printType: 'books',
      };

      if (JSON.stringify(request) === books.request) {
        return;
      }

      setRefreshing(true);
      const { data } = await searchVolumes(request);

      const items = filterByFavorites
        ? data.items?.filter(item => favorites.includes(item.id))
        : data.items;

      dispatch({
        type: 'update',
        total: data.items?.length ?? 0,
        distinctTotal: items.length,
        data: items ?? [],
        page: page,
        request: JSON.stringify(request),
      });

      setRefreshing(false);
    } catch (err) {
      onClearSearch();
      setFilterByFavorites(false);
      Alert.alert('Could not retrieve books.');
    }
  }, [books, favorites, filterByFavorites, orderBy, searchTextDebounce]);

  const onClearSearch = () => dispatch({ type: 'clear' });

  const toggleOrderBy = () =>
    setOrderBy(orderBy === 'newest' ? 'relevance' : 'newest');

  // Toggle the filter, will make table show only favorited books
  const toggleFavoriteFilter = () => setFilterByFavorites(!filterByFavorites);

  // Toggle single book favorited / not favorited
  const toggleFavorite = (identifier: string) => {
    let updatedFavorites;

    if (favorites.includes(identifier)) {
      updatedFavorites = favorites.filter(fav => fav !== identifier);
    } else {
      updatedFavorites = [...favorites, identifier];
    }

    updatedFavorites = updatedFavorites.filter(fav => !!fav);

    setFavorites(updatedFavorites);
    storeFavorites(updatedFavorites);
  };

  const isFavorite = (identifier: string) => favorites.includes(identifier);

  const debounceSearch = useDebounce(setSearchTextDebounce, 500);

  const onChangeSearch = (text: string) => {
    setSearchText(text);
    debounceSearch(text);
  };

  const onChangePage = (page: number) => dispatch({ type: 'moveToPage', page });

  useEffect(() => {
    fetchBooks();
  }, [fetchBooks]);

  useEffect(() => {
    fetchFavorites();
  }, [fetchFavorites]);

  const contextData: BooksContextData = {
    books,
    refreshing,
    searchText,
    onChangeSearch,
    onClearSearch,
    fetchBooks,
    toggleOrderBy,
    toggleFavoriteFilter,
    orderBy,
    filterByFavorites,
    isFavorite,
    toggleFavorite,
    onChangePage,
  };

  return (
    <BooksContext.Provider value={contextData}>
      {children}
    </BooksContext.Provider>
  );
};
