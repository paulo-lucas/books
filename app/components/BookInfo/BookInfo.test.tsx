import 'react-native';
import React from 'react';
import { fireEvent, render, screen, act } from '../../../tests/test-utils';
import { BookInfo } from './';
import { BookDetailsContext } from '@app/contexts/BookDetailsContext';
import { Volume } from '@app/interfaces/volume';

const mockTitle = 'As crônicas de Nárnia';
const mockAuthors = ['Clive Staples Lewis'];
const mockPublishedDate = '2009';
const mockDescription =
  'Este livro apresenta as sete crônicas de Nárnia integralmente, num único volume.';

const book = {
  id: 'I2QvPwAACAAJ',
  volumeInfo: {
    title: mockTitle,
    authors: mockAuthors,
    publishedDate: mockPublishedDate,
    description: mockDescription,
  },
} as Volume;
const isFavorite = jest.fn(() => true);
const toggleFavorite = jest.fn();

const renderBookInfo = (loading = false) =>
  render(
    <BookDetailsContext.Provider
      value={{ book, isFavorite, toggleFavorite, loading }}>
      <BookInfo />
    </BookDetailsContext.Provider>,
  );

describe('book info', () => {
  it('should render correctly', async () => {
    renderBookInfo();
    expect(screen.toJSON()).toMatchSnapshot('book info');
  });

  it('should render skeleton correctly', () => {
    renderBookInfo(true);
    expect(screen.toJSON()).toMatchSnapshot('book info skeleton');
  });

  it('should show book fields', () => {
    renderBookInfo();
    expect(screen.getByText(mockTitle)).toBeTruthy();
    expect(screen.getByText('By ' + mockAuthors.join(', '))).toBeTruthy();
    expect(screen.getByText(mockPublishedDate)).toBeTruthy();

    act(() => {
      fireEvent.press(screen.getByText('Read more'));
    });

    expect(screen.getByText(mockDescription)).toBeTruthy();
  });
});
