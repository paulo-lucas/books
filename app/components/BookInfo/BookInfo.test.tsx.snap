// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book info should render correctly: book info 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "backgroundColor": undefined,
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "paddingVertical": 20,
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "paddingHorizontal": 20,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "width": "100%",
            }
          }
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              Object {
                "backgroundColor": "transparent",
                "overflow": "hidden",
                "position": "relative",
              }
            }
          >
            <Image
              onLoad={[Function]}
              replaceTheme={[Function]}
              source={
                Object {
                  "uri": "https://www.xy.com/images/placeholder.jpg",
                }
              }
              style={
                Object {
                  "aspectRatio": 0.5846153846153846,
                  "bottom": 0,
                  "height": undefined,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": "80%",
                }
              }
              testID="RNE__Image"
              theme={
                Object {
                  "colors": Object {
                    "background": "#ffffff",
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": Object {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": Object {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#00DAC4",
                    "searchBg": "#303337",
                    "secondary": "#2C2A55",
                    "success": "#52c41a",
                    "surface": "#F8F8F8",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                  "mode": "light",
                  "spacing": Object {
                    "lg": 12,
                    "md": 8,
                    "sm": 4,
                    "xl": 24,
                    "xs": 2,
                  },
                }
              }
              transitionDuration={360}
              updateTheme={[Function]}
            />
            <View
              accessibilityElementsHidden={true}
              collapsable={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "opacity": 1,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "aspectRatio": 0.5846153846153846,
                    "backgroundColor": "#bdbdbd",
                    "height": "100%",
                    "justifyContent": "center",
                    "width": "100%",
                  }
                }
                testID="RNE__Image__placeholder"
              >
                <View
                  accessibilityLabel="loading..."
                  accessibilityRole="none"
                  accessible={false}
                  onLayout={[Function]}
                  replaceTheme={[Function]}
                  style={
                    Array [
                      Object {
                        "borderRadius": 2,
                        "overflow": "hidden",
                      },
                      Object {
                        "backgroundColor": "#bdc6cf",
                        "height": 12,
                        "width": "100%",
                      },
                      undefined,
                      Object {
                        "height": "100%",
                        "width": "100%",
                      },
                    ]
                  }
                  testID="RNE__Skeleton"
                  updateTheme={[Function]}
                >
                  <View
                    collapsable={false}
                    style={
                      Object {
                        "backgroundColor": "rgb(166, 178, 190)",
                        "height": "100%",
                        "opacity": 1,
                        "width": "100%",
                      }
                    }
                  />
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "aspectRatio": 0.5846153846153846,
                  "height": undefined,
                  "width": "80%",
                }
              }
              testID="RNE__Image__children__container"
            />
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "#242424",
              "fontSize": 24,
              "fontWeight": "600",
              "marginTop": 20,
            }
          }
        >
          As crônicas de Nárnia
        </Text>
        <Text
          style={
            Object {
              "color": "#242424",
              "fontSize": 14,
            }
          }
        >
          By 
          Clive Staples Lewis
        </Text>
        <Text
          style={
            Object {
              "color": "#242424",
              "fontSize": 14,
            }
          }
        >
          2009
        </Text>
        <Text
          style={
            Object {
              "color": "#242424",
              "fontSize": 18,
              "fontWeight": "500",
              "marginTop": 20,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#242424",
              "fontSize": 18,
              "fontWeight": "500",
              "marginTop": 20,
            }
          }
        >
          Categories
        </Text>
        <Text
          style={
            Object {
              "color": "#242424",
              "fontSize": 18,
              "fontWeight": "500",
              "marginTop": 20,
            }
          }
        >
          Description
        </Text>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            numberOfLines={5}
            style={
              Object {
                "color": "#242424",
                "fontSize": 16,
              }
            }
          >
            Este livro apresenta as sete crônicas de Nárnia integralmente, num único volume.
          </Text>
          <Text
            style={
              Object {
                "color": "#242424",
                "fontSize": 16,
                "fontWeight": "500",
                "textDecorationLine": "underline",
              }
            }
          >
            Read more
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RNCSafeAreaView>
`;

exports[`book info should render skeleton correctly: book info skeleton 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "backgroundColor": undefined,
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "paddingHorizontal": 20,
        "paddingVertical": 20,
      }
    }
  >
    <View
      accessibilityLabel="loading..."
      accessibilityRole="none"
      accessible={false}
      onLayout={[Function]}
      replaceTheme={[Function]}
      style={
        Array [
          Object {
            "borderRadius": 2,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#bdc6cf",
            "height": 12,
            "width": "100%",
          },
          undefined,
          Object {
            "alignSelf": "center",
            "aspectRatio": 0.5846153846153846,
            "height": undefined,
            "width": "80%",
          },
        ]
      }
      testID="RNE__Skeleton"
      updateTheme={[Function]}
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "rgb(166, 178, 190)",
            "height": "100%",
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      accessibilityLabel="loading..."
      accessibilityRole="none"
      accessible={false}
      onLayout={[Function]}
      replaceTheme={[Function]}
      style={
        Array [
          Object {
            "borderRadius": 2,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#bdc6cf",
            "height": 12,
            "width": "100%",
          },
          undefined,
          Object {
            "height": 24,
            "marginTop": 20,
            "width": "85%",
          },
        ]
      }
      testID="RNE__Skeleton"
      updateTheme={[Function]}
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "rgb(166, 178, 190)",
            "height": "100%",
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      accessibilityLabel="loading..."
      accessibilityRole="none"
      accessible={false}
      onLayout={[Function]}
      replaceTheme={[Function]}
      style={
        Array [
          Object {
            "borderRadius": 2,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#bdc6cf",
            "height": 12,
            "width": "100%",
          },
          undefined,
          Object {
            "height": 18,
            "marginBottom": 20,
            "marginTop": 10,
            "width": "60%",
          },
        ]
      }
      testID="RNE__Skeleton"
      updateTheme={[Function]}
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "rgb(166, 178, 190)",
            "height": "100%",
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      accessibilityLabel="loading..."
      accessibilityRole="none"
      accessible={false}
      onLayout={[Function]}
      replaceTheme={[Function]}
      style={
        Array [
          Object {
            "borderRadius": 2,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#bdc6cf",
            "height": 12,
            "width": "100%",
          },
          undefined,
          Object {
            "height": 20,
            "marginTop": 5,
            "width": "100%",
          },
        ]
      }
      testID="RNE__Skeleton"
      updateTheme={[Function]}
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "rgb(166, 178, 190)",
            "height": "100%",
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      accessibilityLabel="loading..."
      accessibilityRole="none"
      accessible={false}
      onLayout={[Function]}
      replaceTheme={[Function]}
      style={
        Array [
          Object {
            "borderRadius": 2,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#bdc6cf",
            "height": 12,
            "width": "100%",
          },
          undefined,
          Object {
            "height": 20,
            "marginTop": 5,
            "width": "100%",
          },
        ]
      }
      testID="RNE__Skeleton"
      updateTheme={[Function]}
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "rgb(166, 178, 190)",
            "height": "100%",
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      accessibilityLabel="loading..."
      accessibilityRole="none"
      accessible={false}
      onLayout={[Function]}
      replaceTheme={[Function]}
      style={
        Array [
          Object {
            "borderRadius": 2,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#bdc6cf",
            "height": 12,
            "width": "100%",
          },
          undefined,
          Object {
            "height": 20,
            "marginTop": 5,
            "width": "100%",
          },
        ]
      }
      testID="RNE__Skeleton"
      updateTheme={[Function]}
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "rgb(166, 178, 190)",
            "height": "100%",
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      accessibilityLabel="loading..."
      accessibilityRole="none"
      accessible={false}
      onLayout={[Function]}
      replaceTheme={[Function]}
      style={
        Array [
          Object {
            "borderRadius": 2,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#bdc6cf",
            "height": 12,
            "width": "100%",
          },
          undefined,
          Object {
            "height": 20,
            "marginTop": 5,
            "width": "100%",
          },
        ]
      }
      testID="RNE__Skeleton"
      updateTheme={[Function]}
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "rgb(166, 178, 190)",
            "height": "100%",
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      accessibilityLabel="loading..."
      accessibilityRole="none"
      accessible={false}
      onLayout={[Function]}
      replaceTheme={[Function]}
      style={
        Array [
          Object {
            "borderRadius": 2,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#bdc6cf",
            "height": 12,
            "width": "100%",
          },
          undefined,
          Object {
            "height": 20,
            "marginTop": 5,
            "width": "100%",
          },
        ]
      }
      testID="RNE__Skeleton"
      updateTheme={[Function]}
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "rgb(166, 178, 190)",
            "height": "100%",
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </View>
    <View
      accessibilityLabel="loading..."
      accessibilityRole="none"
      accessible={false}
      onLayout={[Function]}
      replaceTheme={[Function]}
      style={
        Array [
          Object {
            "borderRadius": 2,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#bdc6cf",
            "height": 12,
            "width": "100%",
          },
          undefined,
          Object {
            "height": 20,
            "marginTop": 5,
            "width": "70%",
          },
        ]
      }
      testID="RNE__Skeleton"
      updateTheme={[Function]}
    >
      <View
        collapsable={false}
        style={
          Object {
            "backgroundColor": "rgb(166, 178, 190)",
            "height": "100%",
            "opacity": 1,
            "width": "100%",
          }
        }
      />
    </View>
  </View>
</RNCSafeAreaView>
`;
